src/events/broker.py

```diff
@@ -0,0 +1,328 @@
+        )
+        self.tls_manager = TLSManager(
+            state=self.charm.state, workload=self.workload, substrate=self.charm.substrate
+        )
+        self.auth_manager = AuthManager(
+            state=self.charm.state,
+            workload=self.workload,
+            kafka_opts=self.config_manager.kafka_opts,
+            log4j_opts=self.config_manager.tools_log4j_opts,
+        )
+
+        self.balancer_manager = BalancerManager(self)
+
+        self.framework.observe(getattr(self.charm.on, "install"), self._on_install)
+        self.framework.observe(getattr(self.charm.on, "start"), self._on_start)
+        self.framework.observe(getattr(self.charm.on, "config_changed"), self._on_config_changed)
+        self.framework.observe(getattr(self.charm.on, "update_status"), self._on_update_status)
+        self.framework.observe(getattr(self.charm.on, "secret_changed"), self._on_secret_changed)
+
+        self.framework.observe(self.charm.on[PEER].relation_changed, self._on_config_changed)
+
+        self.framework.observe(
+            getattr(self.charm.on, "data_storage_attached"), self._on_storage_attached
+        )
+        self.framework.observe(
+            getattr(self.charm.on, "data_storage_detaching"), self._on_storage_detaching
+        )
+
+    def _on_install(self, _: InstallEvent) -> None:
+        """Handler for `install` event."""
+        self.config_manager.set_environment()
+        self.charm.unit.set_workload_version(self.workload.get_version())
+
+    def _on_start(self, event: StartEvent) -> None:
+        """Handler for `start` event."""
+        if self.charm.state.peer_relation:
+            self.charm.state.unit_broker.update({"cores": str(self.balancer_manager.cores)})
+
+        self.charm._set_status(self.charm.state.ready_to_start)
+        if not isinstance(self.charm.unit.status, ActiveStatus):
+            event.defer()
+            return
+
+        # required settings given zookeeper connection config has been created
+        self.config_manager.set_zk_jaas_config()
+        self.config_manager.set_server_properties()
+        self.config_manager.set_client_properties()
+
+        # start kafka service
+        self.workload.start()
+        logger.info("Kafka snap started")
+
+        # check for connection
+        self.charm.on.update_status.emit()
+
+        # only log once on successful 'on-start' run
+        if isinstance(self.charm.unit.status, ActiveStatus):
+            logger.info(f'Broker {self.charm.unit.name.split("/")[1]} connected')
+
+    def _on_config_changed(self, event: EventBase) -> None:
+        """Generic handler for most `config_changed` events across relations."""
+        if self.charm.state.balancer and self.charm.unit.is_leader():
+            self.charm.state.balancer.update(
+                {
+                    "broker-capacities": self.config_manager.broker_capacities,
+                    "rack_aware": json.dumps(bool(self.config_manager.rack_properties)),
```

Comment:
You are absolutely right, this is an error since the relation field we fetch does use the hyphen. The only reason it was not caught until now is that we are not doing anything with it. Changed in 9bef8c5a8e377ae088a0304a93192b094c40e056